aten/src/ATen/function_wrapper.py

```diff
@@ -99,7 +101,8 @@ def TypedDict(name, attrs, total=True):  # type: ignore
 """)
 
 DEFAULT_FUNCTION_REGISTRATION = CodeTemplate("""\
-m.impl("${unqual_operator_name_with_overload}", TORCH_FN(TypeDefault::${type_wrapper_name}));
+m.impl("${unqual_operator_name_with_overload}",
+       c10::impl::hacky_wrapper_for_legacy_signatures(TORCH_FN(TypeDefault::${type_wrapper_name})));
```

Comment:
It's a bit of a bummer that our nice clean API has so quickly gotten encrusted with this shim, in particular to the extent these calls will serve as example API usage. Ergonomically it's kind of rough (especially since it might do nothing). Two ideas:

1. teach the codegen to only include the wrapper when necessary. I realize this intrudes on the complete decoupling you were going for here, but I think it would be worth it - the API usage will be better exemplified both for legacy and non-legacy cases. I.e., I think the value of complete decoupling is outweighed by the extent to which it obscures the (evolving) relationship between registered and implemented signatures, and by the readability hit. 

To pitch the change another way: we want registrations for kernels that are already in our desired final state to look like it.

2. hide the wrapper behind TORCH_FN, or a macro that wraps TORCH_FN. I think this is less desirable since it simply hides everything, rather than clarifying when the wrapper is needed, but it would fix the readability issue.

We can teach the codegen when to emit the wrapper just by looking for tensor options in the signature, correct? If so, I'd suggest having a function `needs_legacy_wrapper` where any future triggering conditions would also go (per the comment header in above the metaprogramming for this)